Core: Fix monorepo compatibility

merceyz opened the pull request:
## Issue

Storybook doesn't work correctly in monorepos with both node_modules (Yarn 1/2) and PnP (Yarn 2) because it's trying to require the addons declared in `.storybook/main.js` from `@storybook/core` instead of requiring them on behalf of the config.

The e2e tests currently pass because Yarn has a compatability feature, called `fallbackMode`, that allows dependencies to require top-level dependencies without declaring them as their own. In a monorepo this fails as the top level no longer declares these dependencies, the workspace does, and under the node_modules linker the addons are not guaranteed to be hoisted to the same level as `@storybook/core`.

## What I did

- Disabled the `fallbackMode` to highlight the issue in the e2e tests https://yarnpkg.com/configuration/yarnrc#pnpFallbackMode
- Resolve preset relative to config
- @gaetanmaisse - Fix issue with preset name checks as they are now resolved path and no more simple names. 

## How to test

Disable the `fallbackMode` and try to use storybook in a PnP project

## Extra

cc @gaetanmaisse since you worked on PnP compatability

merceyz committed:
ci: disable fallback mode

merceyz committed:
test: update pnp test

merceyz committed:
chore: simpllify test setup

gaetanmaisse commented:
Thanks @merceyz, this looks interesting! I will deep dive in this when I will come back from holidays in 10 days.

merceyz committed:
fix: avoid overwriting dev config

merceyz committed:
fix: resolve relative to config

merceyz committed:
ci: add package extensions

merceyz committed:
Revert "fix: avoid overwriting dev config"

This reverts commit 83831a58cd33c4d100ae1e0870b683809a11f99e.

merceyz commented on lib/core/src/server/presets.js:
```diff
@@ -69,7 +69,10 @@ export const resolveAddonName = (configDir, name) => {
 
-  return { name, type: 'presets' };
+  return {
+    name: resolveFrom(configDir, name),
```
This fixes the issue with dependencies not getting required on behalf of the config, however it breaks other things (see CI output) and i'm not sure why. @gaetanmaisse do you have any suggestions?

gaetanmaisse replied:
I'm still AFK for 2 days but this is the first PR I will take a look on Monday ;)

gaetanmaisse replied:
🤔 , interesting, CI is failing for CRA tests only, for all the other setups it looks ok. @ndelangen I think I will need to discuss with you on this one.

ndelangen replied:
@gaetanmaisse sure, I'm back from vacation starting today, hit me up on discord?

ndelangen replied:
@gaetanmaisse let's go! https://calendly.com/ndelangen/storybook

merceyz committed:
Merge branch 'next' into pnp

merceyz committed:
chore: revert leftover merge changes

gaetanmaisse committed:
fix: resolve presets relatively to configDir

Storybook doesn't work correctly in monorepos with both node_modules (Yarn 1/2) and PnP (Yarn 2) because it's trying to require the addons declared in .storybook/main.js from @storybook/core instead of requiring them on behalf of the config.

Also makes checks on preset names less restrictive as they are now resolved path and not simple name anymore.

merceyz committed:
chore: fix test

merceyz committed:
fix: handle windows paths

gaetanmaisse committed:
chore: fix tests after changes in `resolveAddonName` behavior

gaetanmaisse committed:
Merge branch 'next' into pnp

gaetanmaisse commented:
@merceyz All green! 🎉  Do you have something to add or we can open the PR?

merceyz commented:
@gaetanmaisse Nope, all good, thanks for finding the issue with the CRA preset :tada:

merceyz commented:
Note to whomever merges this: You should use Squash and Merge since the commit history is a bit messy here

shilman commented:
@merceyz no worries about the commit history. we don't use squash merges in this project! 😁

merceyz commented:
> we don't use squash merges in this project

That makes the commit history really hard to read but alright

khats commented:
How to resolve preset relative to storybook installed folder, but not config?
there is following issue after resolving preset relative to config:
storybook installed under `./build/storybook` . one of monorepo packages locates in `./monorepo-pack1` and has `./monorepo-pack1/./storybook/main.js` with enabled addon: `@storybook/addon-essentials` . There is symlink to `start-storybook` from `./monorepo-pack1` to `./build/storybook/node_modules/@storybook/react` . 

After changes in this pull request  storybook tries to look `./monorepo-pack1/node_modules/@storybook/addon-essentials`, but expected for me is `./build/storybook/node_modules/@storybook/addon-essentials` where is storybook and other addons installed. 
I want to have different configurations for monorepo for each sub project, but don't want to add `@storybook/addon-essentials` for each project as devDependency

merceyz commented:
Resolve the presets upfront yourself using `require.resolve`

khats commented:
@merceyz thank your for answer
After changing addon name  in `main.js` from `@storybook/addon-essentials` to   `require.resolve(path.join(__dirname,"../../build/storybook/node_modules/@storybook/addon-essentials"))` there is no more error `Addon value should end in /register OR it should be a valid preset` and addon(viewports) works, but i have warning 
`Expected '@storybook/addon-docs' to be listed before '@storybook/addon-controls' (or '@storybook/addon-essentials'). Check your main.js? 
https://github.com/storybookjs/storybook/issues/11442
`

khats commented:
in case of using addon-essentials with custom full path like: `require.resolve("../..build/storybook/node_modules/@storybook/addon-essentials")` `ensureDocsBeforeControls` fails because expects that name of the addon will start with "@storybook/addon-essentials". Looks like fix could be not using `name.startsWith(addon)` for finding addon, but `name.includes(addon)`

upd: i have found same implementation already done
https://github.com/storybookjs/storybook/commit/d5efc500488e6e7393fe59745b317e1c2c70b024
